src/main/java/org/oscii/concordance/Word2VecManager.java

```diff
@@ -0,0 +1,262 @@
+   *
+   * @param lang
+   * @param query
+   */
+  public boolean containsQuery(String lang, String query) {
+    return supports(lang) && models.get(lang).forSearch().contains(query);
+  }
+
+  /**
+   * Checks whether the degraded query (e.g. lowercased) is part of
+   * model for given language.
+   *
+   * @param lang
+   * @param query
+   */
+  public boolean containsDegradedQuery(String lang, String query) {
+    return supports(lang) && models.get(lang).forSearch().contains(Lexicon.degrade(query));
+  }
+
+  /**
+   * Returns the query or degraded query if it is part of the model
+   * for a given language.
+   *
+   * @param lang
+   * @param query
+   */
+  public String getMatchingQuery(String lang, String query) {
+    if (containsQuery(lang, query)) {
+      return query;
+    } else if (containsDegradedQuery(lang, query)) {
+      return Lexicon.degrade(query);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Ranks a list of cocordances by calculating the similarity of each
+   * example's mean vector to the mean vector of the context. The
+   * result is sorted according to distance/similarity. In order to
+   * reduce computation time, the candidates are reduced in size by
+   * filtering out short entries (likely stop words) and truncating
+   * the result to a maximum number of entries.
+   *
+   * @param lang
+   * @param context
+   * @param concordances
+   */
+  public boolean rankConcordances(String lang, String context, List<SentenceExample> concordances) {
+    if (!supports(lang) || context.length() == 0) {
+      return false;
+    }
+    Searcher searcher = models.get(lang).forSearch();
+    // retrieve and score context
+    String[] contextTokens = reduceTokens(context.split("\\s+"), MIN_SEG_LEN, MIN_TOK_LEN, MAX_RES_LEN);
+    double[] contextMean = searcher.getMean(contextTokens);
+    // iterate over concordance results
+    concordances.forEach(ex -> {
+        String[] tokens = reduceTokens(ex.sentence.tokens, MIN_SEG_LEN, MIN_TOK_LEN, MAX_RES_LEN);
+        double[] tokensMean = searcher.getMean(tokens);
+        logger.debug("tokens={} context={} ({})", tokens, contextTokens, contextTokens.length);
+        logger.debug("means: tokens=[{},{},{},...] context=[{},{},{},...]",
+                     tokensMean[0], tokensMean[1], tokensMean[2],
+                     contextMean[0], contextMean[1], contextMean[2]);
+        double dist = searcher.cosineDistance(tokensMean, contextMean);
+        ex.similarity = (Double.isNaN(dist) ? -1.0 : dist);
+        logger.debug("distance: {}", ex.similarity);
+      });
+    // rerank according to word2vec similarities
+    Collections.sort(concordances, Order.bySimilarity);
+    return true;
+  }
+
+  /**
+   * Returns the raw word vector (n-dimensional array of doubles) for
+   * given query and language.
+   *
+   * @param lang
+   * @param query
+   */
+  public List<Double> getRawVector(String lang, String query) throws UnknownWordException {
```

Comment:
no inner loop served from this, it's just for the service request `wordvec=true` which serves the raw vector for a query